subsystems/DrivetrainBase.java

```diff
@@ -1,51 +1,112 @@
+import edu.wpi.first.wpilibj.Encoder;
 import edu.wpi.first.wpilibj.SpeedControllerGroup;
+import edu.wpi.first.wpilibj.Timer;
 import edu.wpi.first.wpilibj.controller.PIDController;
 import edu.wpi.first.wpilibj2.command.PIDSubsystem;
 
 public abstract class DrivetrainBase extends PIDSubsystem {
-    // create leader and follower motors for the drivetrain
+    // create leader and follower motors for the drivetrain, as well as encoders that will attach to the leaders.
     protected WPI_TalonFX leftLeader, rightLeader, leftFollower, rightFollower;
+    protected Encoder leftEncoder, rightEncoder;
 
     // group the leader and follower motors so they can be controlled at the same time
     protected SpeedControllerGroup leftGroup, rightGroup;
 
     // make the speed controller groups into one drivetrain object
     protected DifferentialDrive differentialDrive;
 
-    // create variables needed for PID. these will all be changed by the subclass.
+    // create a hashtable to look up constants
+    protected Hashtable<String, Object> constants;
+
+    // create variables needed for odometry.
     protected AHRS ahrs;
     protected FPID turnFPID;
     protected double turnFPIDThrottle;
     protected boolean invertGyro;
-
-    // placeholders for a SparkMax-based drivetrain
-    /*
-    protected CANSparkMax leftLeader = new CANSparkMax(0, MotorType.kBrushless);
-    protected CANSparkMax rightLeader = new CANSparkMax(1, MotorType.kBrushless);
-    protected CANSparkMax leftFollower = new CANSparkMax(2, MotorType.kBrushless);
-    protected CANSparkMax rightFollower = new CANSparkMax(3, MotorType.kBrushless);
-    */
+    protected int invertLeft;
+    protected int invertRight;
+    protected DifferentialDriveOdometry odometry;
+    private final LimitedPoseMap poseHistory;
+    private double lastAngleSetpoint = Double.NaN;
+    private final PIDController turnController;
 
     // set all of the variables from the subclass to this abstract class
-    public DrivetrainBase(WPI_TalonFX leftLeader, WPI_TalonFX rightLeader, WPI_TalonFX leftFollower, WPI_TalonFX rightFollower, FPID turnFPID, double turnFPIDThrottle, boolean invertGyro) {
-        super(new PIDController(turnFPID.kP, turnFPID.kI, turnFPID.kD));
-        
-        this.leftLeader = leftLeader;
-        this.rightLeader = rightLeader;
-        this.leftFollower = leftFollower;
-        this.rightFollower = rightFollower;
-        this.turnFPID = turnFPID;
-        this.turnFPIDThrottle = turnFPIDThrottle;
-        this.invertGyro = invertGyro;
-        leftGroup = new SpeedControllerGroup(leftLeader, leftFollower);
-        rightGroup = new SpeedControllerGroup(rightLeader, rightFollower);
+    public DrivetrainBase(Hashtable<String, Object> constants, WPI_TalonFX...motors) {
+        super(new PIDController((double) ((FPID) constants.get("TURN_FPID")).kP, (double) ((FPID) constants.get("TURN_FPID")).kI, (double) ((FPID) constants.get("TURN_FPID")).kD));        
```

Comment:
I'm just saying, I think an object with defined values may be easier to work with than a hash map. Then you won't have to go through this mess of making everything an `Object` in the hash map, and then casting it back into its original form.